### PR TITLE
fix salt subset in orchestrator

### LINKED ISSUE #46456: "ValueError" when running orch with "subset"
### Description of Issue/Question

When running an orch with "subset: 1" (or any other number) I get "```ValueError: dictionary update sequence element #0 has length 1; 2 is required```" Python exception.

Removing "subset" line make orch run as expected.

### Setup
my-orch.sls

```
problem:
  salt.state:
    - tgt: '*'
    - sls: my.state.name
    - subset: 1
```

### Steps to Reproduce Issue

```
$ sudo salt-run state.orch my-orch

[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/state.py", line 1851, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1795, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/states/saltmod.py", line 272, in state
    cmd_ret = __salt__['saltutil.cmd'](tgt, fun, **cmd_kw)
  File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1322, in cmd
    client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1287, in _exec
    fcn_ret.update(ret_comp)
ValueError: dictionary update sequence element #0 has length 1; 2 is required

my-vm.example.com_master:
----------
          ID: problem
    Function: salt.state
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1851, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1795, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/states/saltmod.py", line 272, in state
                  cmd_ret = __salt__['saltutil.cmd'](tgt, fun, **cmd_kw)
                File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1322, in cmd
                  client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1287, in _exec
                  fcn_ret.update(ret_comp)
              ValueError: dictionary update sequence element #0 has length 1; 2 is required
     Started: 03:38:45.070976
    Duration: 2349.035 ms
     Changes:

Summary for my-vm.example.com_master
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   2.349 s
```

We debugged this a bit and the exception is in this loop in modules/saltutil.py:

```
    for ret_comp in _cmd(**cmd_kwargs):
       	fcn_ret.update(ret_comp)
```

Normally, _cmd() here returns a generator that returns nested structures with a jid and another attribute holding returned data. But when using "subset", _cmd() returns a dictionary with a single minion id as key and return data as value, so "for" then iterates over dictionary keys and thus .update() blows up trying to update fcn_ret dict with a string value.

### Versions Report

```
Salt Version:
           Salt: 2017.7.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.8.1
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.5
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.10 (default, Dec 19 2015, 08:44:39)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.11.6.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core
```

### PR DESCRIPTION
**Pull request by gtmanfred**
### What does this PR do?
This was originally supposed to be part of https://github.com/saltstack/salt/pull/43933

But went missing for some reason, possibly in my rebase to add the comment about popping expect_minions

### What issues does this PR fix or reference?
Fixes #46456

### Tests?

Yes

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/modules/saltutil.py [hunk 1]
```diff
@@ -1273,7 +1273,7 @@ def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs):
         _cmd = client.cmd_subset
         cmd_kwargs = {
             'tgt': tgt, 'fun': fun, 'arg': arg, 'tgt_type': tgt_type,
-            'ret': ret, 'kwarg': kwarg, 'sub': kwargs['subset'],
+            'ret': ret, 'cli': True, 'kwarg': kwarg, 'sub': kwargs['subset'],
```
Comment (cachedout) [inline salt/modules/saltutil.py]: I don't understand why this is the correct fix. Could you explain in more detail?

Comment (gtmanfred) [reply salt/modules/saltutil.py]: This is how subset was originally implemented for the `salt` command line

https://github.com/saltstack/salt/commit/27cc8295e69

So calling it from the saltutil command options need to call it the same way.

https://github.com/saltstack/salt/blob/2017.7/salt/cli/salt.py#L144

Comment (cachedout) [reply salt/modules/saltutil.py]: Can we either refactor this then so there is a single source of truth for this behavior or heavily document in both places why this behavior is necessary? When we have these mystery options being passed with no clear documentation on why they are necessary or what they do, it feels a little tech debt-y to me. :]

Comment (gtmanfred) [reply salt/modules/saltutil.py]: :+1:

**Comment by rallytime** [comment]
FYI @gtmanfred - just a bump on this one. :)

File: salt/client/__init__.py [hunk 2]
```diff
@@ -481,6 +481,8 @@ def cmd_subset(
         following exceptions.
 
         :param sub: The number of systems to execute on
+        :param cli: When this is set to True, a generator is returned,
```
[the remark this answers is not in the document]
Comment (gtmanfred) [reply salt/client/__init__.py]: It isn't, but I think we should change it in develop and deprecate it going forward from there.

I can make an issue for that if you want?

**Pull request merged**